File: python/0680-e-valid-palindrome-ii.py
## https://leetcode.com/problems/valid-palindrome-ii

class Solution:
    def validPalindrome(self, s: str) -> bool:
        mid = len(s)//2 + 1
        reversed_half = s[-mid:][::-1]

        if s[:mid] == reversed_half:
            return True

        for i in range (0, mid):
            if s[i] != reversed_half[i]:
                skip_left = s[i+1:mid] == reversed_half[i:mid-1]
                skip_right = s[i:mid-1] == reversed_half[i+1:]
                return skip_left or skip_right

        return True

if __name__ == '__main__':
    print(Solution().validPalindrome("aba"))
    print(Solution().validPalindrome("abca"))
    print(Solution().validPalindrome("abbac"))
    print(Solution().validPalindrome("lccul"))
    print(Solution().validPalindrome("acxcybycxcxa"))
    print(Solution().validPalindrome("eeccccbebaeeabebccceea"))
    print(Solution().validPalindrome("eeccccbebaeeabebcccee"))

File: kotlin/0680-e-valid-palindrome-ii.kt
// https://leetcode.com/problems/valid-palindrome-ii
class Solution {
    fun validPalindrome(s: String): Boolean {
        var l = 0
        var r = s.length - 1

        while (l < r) {
            if (s[l] == s[r]) {
                l++
                r--
            } else return validPalindrome(s, l+1, r) || validPalindrome(s, l, r-1)
        }

        return true
    }

    fun validPalindrome(s: String, l: Int, r: Int): Boolean {
        var l = l
        var r = r

        while (l < r) {
            if (s[l] == s[r]) {
                l++
                r--
            } else return false
        }

        return true
    }
}